src/MySqlConnector/Core/ICancellableCommand.cs

```diff
@@ -53,28 +54,41 @@ public static int GetCommandTimeUntilCanceled(this ICancellableCommand command)
 	/// method call.</remarks>
 	public static void ResetCommandTimeout(this ICancellableCommand command)
 	{
+		// read value cached on the command
+		var effectiveCommandTimeout = command.EffectiveCommandTimeout;
+
+		// early out if there is no timeout
+		if (effectiveCommandTimeout == Constants.InfiniteTimeout)
+			return;
+
 		var session = command.Connection?.Session;
-		if (session is not null)
+		if (session is null)
+			return;
+
+		// determine the effective command timeout if not already cached
+		if (effectiveCommandTimeout is null)
+		{
+			effectiveCommandTimeout = command.CommandTimeout == 0 || session.CancellationTimeout == 0 ?
+				Constants.InfiniteTimeout :
+				GetCommandTimeUntilCanceled(command);
```

Comment:
It might be interesting to completly remove GetCommandTimeUntilCanceled method : it is used only there, and will do another test on commandTimeout being equal to 0, and session existing, while this has just been tested here. 

second thing is this : 
CancellationTimeout is defined to be 

> The length of time (in  seconds) to wait for a query to be canceled when [CommandTimeout](https://mysqlconnector.net/api/mysqlconnector/MySqlCommand/CommandTimeout/) expires, or zero for no timeout.

so why having this 
```
return Math.Min(commandTimeout, Math.Max(1, (int.MaxValue / 1000) - Math.Max(0, session.CancellationTimeout)));
```
in place of just using `commandTimeout`

Reply:
Yes, I should inline that method rather than just making it `private`.

The `Math.Min` is to avoid integer overflow edge cases. For example, let's say the user set `CommandTimeout = 2_500_000` and `CancellationTimeout = 10`. 

Because these have to be multiplied by 1,000 (to convert to milliseconds), it would integer overflow to -1,794,957,296. Thus, we clamp the `commandTimeout` value to 2,147,483 - CancellationTimeout (= 2,147,473 in this example) so we don't overflow later when multiplying by 1000.